keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,865 @@
+matchConditions:
+  # Exclude node requests from the webhook
+  - expression: '!("system:nodes" in request.userInfo.groups)'
+```
+
+#### Scope an NFS access management webhook to Pods mounting NFS volumes
+
+> I want to narrowly scope my webhook to only the relevant requests, in order to reduce load on the
+> webhook and reduce latency in irrelevant requests.
+
+Concrete example:
+
+> A NFS deployment uses an third-party access management system. I have an admission webhook that
+> performs an access check for against the external system for pods that mount NFS volumes. Only
+> pods with NFS volumes need to be checked.
+
+Currently, there is no way to achieve this. Many webhook implementations today start by checking
+that the request is within scope, and return early if it's not. This adds latency and an additional
+failure point to irrelevant requests. This example requires an external integration, and thus is not
+a candidate for migration to CEL `ValidatingAdmissionPolicy`.
+
+With match conditions, the expressions can check whether the request object is in-scope for the
+webhook:
+
+```yaml
+rules:
+  - operations: ['CREATE']
+    apiGroups: '' # core
+    apiVersions: '*'
+    resources: 'pods'
+matchConditions:
+  # Only include pods with an NFS volume.
+  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
+```
+
+### Goals
+
+1. Provide a filtering mechanism for excluding requests from an admission webhook
+2. Maintain consistency with `ValidatingAdmissionPolicy`
+
+### Non-Goals
+
+* Provide a mechanism to exclude requests from all webhooks.
+
+## Proposal
+
+### API
+
+Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
+with a new `MatchConditions` field:
+
+```go
+
+type ValidatingWebhook struct {
+  // ...
+
+  // MatchConditions is a list of conditions on the AdmissionRequest ('request') that must be met for a
+  // request to be sent to this webhook.
+  // +optional
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+type MutatingWebhook struct {
+  // ...
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+// MatchCondition represents a condition which must by fulfilled for a request to be sent to a webhook.
+type MatchCondition struct {
+	// Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the AdmissionRequest, organized into CEL variables:
+	//
+	//'object' - The object from the incoming request. The value is null for DELETE requests.
+	//'oldObject' - The existing object. The value is null for CREATE requests.
+	//'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// request.object. No other metadata properties are accessible.
```

Comment:
above it says `object`, here it says `request.object`.  Which is correct?

Reply:
My understanding is that `object` is just a convenience alias to `request.object`, but I changed this to `object` for consistency.

Reply:
For ease of use in ValidationAdmissionPolicy, we _moved_ `object` out of request top level (`request.object` is not allowed). We did the same for `oldObject`. It should be trivial to use the same bit of code here so that `object` / `oldObject` work exactly the same in both APIs.

Reply:
> For ease of use in ValidationAdmissionPolicy, we moved object from request to top level (request.object is not allowed).

This would work for me.  Allowing both seems strange to me.

Reply:
IMO it would be simpler to leave the `request` object alone, and make `object` an alias (or copy, if CEL doesn't have that notion). That way I can just look at the API documentation for `AdmissionReview` to know what fields are available.

Otherwise, the documentation will need to say something like "request is basically an AdmissionReview struct, with the following differences ..."

Reply:
I honestly don't feel super strong about if object should be top level or not.  I'm less excited about having two ways to access the same variable (object and request.object). Maybe more pressing is that whatever we decide to do here is kept in sync with ValidationAdmissionPolicy (either by changing ValidationAdmissionPolicy or conforming to what it does).